docs/SPEC.md

```diff
@@ -3685,31 +3685,39 @@ from(bucket: "telegraf/autogen")
-
-TODO(nathanielc): The fieldFn is not valid and needs to change. It uses dynamic object keys which is not allowed.
+The `to` operation takes data from a stream and writes it to a bucket.
+`to` has the following properties:
+
+| Name              | Type           | Description                                                                                                                                                                                                                        |
+| ----              | ----           | -----------                                                                                                                                                                                                                        |
+| bucket            | string         | Bucket is the bucket name into which data will be written.                                                                                                                                                                         |
+| bucketID          | string         | BucketID is the bucket ID into which data will be written.                                                                                                                                                                         |
+| org               | string         | Org is the organization name of the bucket.                                                                                                                                                                                        |
+| orgID             | string         | OrgID is the organization ID of the bucket.                                                                                                                                                                                        |
+| host              | string         | Host is the location of a remote host to write to. Defaults to `""`.                                                                                                                                                               |
+| token             | string         | Token is the authorization token to use when writing to a remote host. Defaults to `""`.                                                                                                                                           |
+| measurementColumn | string         | MeasurementColumn is the measurement into which data will be written. Defaults to `"_measurement"`.                                                                                                                                |
+| fieldColumn       | string         | FieldColumn defines the column from which the field keys and values will be taken.  Defaults to `"_field"`.                                                                                                                                   |
```

Comment:
in general, the fieldFn is not a useful construct anymore since we can compute a new field using `with` in a map call.  So it makes sense to drop it.   However, should we allow the possibility for the field name to be the column name?  because that's how it would come out if we used `map`.  there's not a convenient way to 'unpivot' back to a _field column.

Reply:
Unpivoting is simple using `map`,

```
|> map(fn:(r) => ({_field : "max", _value: r.max}))
```

This only works for a single value, if you have multiple values we would need a `flatMap` function. 


```
|> flatMap(fn:(r) => [{_field : "max", _value: r.max}, {_field: "min", _value: r.min}])
```

Reply:
Nice, `flatMap` would be a great addition.

@adam I think that within a single function it's tough to satisfy both the trivial case of `from |> range |> to` and to also allow more intuitively structured data (e.g., value column label becomes field key in influxdb) to work right.

Currently no one is using `fieldFn` and `tagColumns` in production, so the former case is all that we need to support.

Also, it's worth noting that when we have a working `flatMap` function, we could wrap this simple `to` with another function to handle data that has been pivoted.

Reply:
nobody is using fieldFn and tagColumns in production because it has an automagic default routine that happens to be inexpressible in the flux language syntax.  The current behavior that I expect most users will be surprised if it's lost is: 

_measurement == measurement name
field names:  all field names returned by fieldFn, which like you said is pretty much always whatever is in _field
tags: everything that's not a field column.  

With flatmap, you're looking at a headache of lookups from a flatmap:  _fieldNames1, _fieldNames2, ..., _value1, _value2, ....

I just want to make sure we like how that would work out.  It might be more future-proof to take that kind of data as input and pivot it.  Regardless, if we want to be prepared for this case then the field identifiers should be lists:  fieldColumns: ["_fieldNames1", "_fieldNames2", ...] and similarly for valueColumn(s)

Reply:
also, consider you are proposing this to perform a pivot in-line with the writes.  this is not a deal breaker, but it also is only half true to say "we don't have to pivot the data"